File: problemset/Simplify_Path.py
class Solution(object):
    def simplifyPath(self, path):
        """
        :type path: str
        :rtype: str
        """
        parts = path.strip().strip("/").split("/")
        stack = []
        for p in parts:
            if p == "." or not p: continue
            elif p == "..": stack.pop(len(stack)-1) if stack else None
            else: stack.append(p)
        return "/"+"/".join(stack)

File: problemset/Simplify_Path.cpp
class Solution {
public:
    string simplifyPath(string path) {
      int n = path.length();
      if (!n) return "";
      vector<string> stk;
      for (int i = 0, j = 0; i < n; i = j) {
        assert(path[i] == '/');
        j = i+1;
        while (j < n && path[j] != '/') ++j;
        int len = j - 1 - i;
        if (len == 0 || (len == 1 && path[i+1] == '.')) {
          continue;
        } else if (len == 2 && path[i+1] == '.' && path[i+2] == '.') {
          if (stk.size()) stk.pop_back();
          continue;
        } else {
          string block = path.substr(i+1, len);
          stk.push_back(block);
        }
      }
      string ret("/");
      for (int i = 0; i < stk.size(); ++i) {
        ret += stk[i];
        if (i != stk.size()-1) ret += "/";
      }
      return ret;
    }
};
